src/Components/Web.JS/src/Platform/Circuits/CircuitManager.ts

```diff
@@ -399,3 +547,87 @@ export class CircuitManager implements DotNet.DotNetCallDispatcher {
     }
   }
 }
+
+class CircuitState<T> {
+
+  public constructor(
+    private _stateName: string,
+    _initialValue?: T,
+    private _resetValue?: T
+  ) {
+    this._lastValue = _initialValue;
+  }
+
+  private _promise?: Promise<T>;
+
+  private _resolve?: (value: T | PromiseLike<T>) => void;
+
+  private _reject?: (reason: any) => void;
+
+  private _lastValue?: T;
+
+  public reset(): void {
+    if (this._promise) {
+      throw new Error(`Circuit state ${this._stateName} is already in progress`);
+    }
+
+    const { promise, resolve, reject } = Promise.withResolvers<T>();
+    this._promise = promise;
+    this._resolve = resolve;
+    this._reject = reject;
+    this._lastValue = this._resetValue;
+  }
+
+  public complete(value: T): void {
+    if (!this._resolve) {
+      throw new Error(`Circuit state ${this._stateName} not initialized`);
+    }
+
+    const resolve = this._resolve;
+    this._lastValue = value;
+
+    this._promise = undefined;
+    this._resolve = undefined;
+    this._reject = undefined;
+
+    resolve(value);
+  }
+
+  public fail(reason: any): void {
+    if (!this._reject) {
+      throw new Error(`Circuit state ${this._stateName} not initialized`);
+    }
+
+    const reject = this._reject;
+
+    this._promise = undefined;
+    this._resolve = undefined;
+    this._reject = undefined;
+
+    reject(reason);
+  }
+
+  public isInprogress(): boolean {
+    return !!this._promise;
+  }
+
+  public currentProgress(): Promise<T> {
+    if (!this.isInprogress()) {
+      throw new Error(`Circuit state ${this._stateName} is not in progress`);
+    }
+
+    return this._promise!;
+  }
+
+  public transitionTo(newState: T): void {
+    if (this._promise) {
```

Comment:
are there any invalid transitions on the graph ?

Reply:
Nope.

This helper class roughly encapsulates the transitions to make thins "saner" when we set them up in the CircuitManager. The `CircuitManager` takes care of transitioning the right way.